pulpcore/app/tasks/export.py

```diff
@@ -58,10 +60,22 @@ def _export_to_file_system(path, content_artifacts, method=FS_EXPORT_METHODS.WRI
     ):
         raise RuntimeError(_("Only write is supported for non-filesystem storage."))
 
+    publication_paths = {}
+    if publication:
+        publication_paths = {
+            content_artifact_pk: relative_path
+            for content_artifact_pk, relative_path in publication.published_artifact.values_list(
+                "content_artifact__pk", "relative_path"
+            )
+        }
+
     for ca in content_artifacts.select_related("artifact").iterator():
         artifact = ca.artifact
-        dest = os.path.join(path, ca.relative_path)
+        relative_path = ca.relative_path
+        if publication and ca.published_artifact:
```

Comment:
This is do not understand. And maybe it's a deficiency in our data model. I would expect that the same content artifact could be published as part of a multitude of different publications and therefore be associated with multiple published artifacts.
Also a publication (that is not pass-through) may even decide to not publish some content artifacts at all. Should we not derive the list of relative paths and their artifacts from the query above if a publication is provided?

Reply:
> I would expect that the same content artifact could be published as part of a multitude of different publications and therefore be associated with multiple published artifacts.

That is correct. Same content artifact may be associated to multiple publications. I should probably just remove the `ca.published_artifact` part. The idea is this method is given the correct publication to use. So may be a change like below should be sufficient?

```suggestion
        if publication:
```



> Also a publication (that is not pass-through) may even decide to not publish some content artifacts at all. Should we not derive the list of relative paths and their artifacts from the query above if a publication is provided?
Good point I dont know if we can do a fs_export of a pass through publication. May be I should change it to

 ```suggestion
        if publication and not publication.pass_through:
```

Reply:
updated

Reply:
Digging this up from memory: A pass through publication serves all content artifacts in the repo-version according to it's relative-path (silently assuming there is one). But also some extra published artifacts (usually metadata).
So I think, you want to include all the published artifacts if there is a publication and all the content artifacts if there is none or it is pass through. Right?
As far as I understand, the goal is to create a tar representation of the looks of the publication on the webserver.

Reply:
> So I think, you want to include all the published artifacts if there is a publication and all the content artifacts if there is none or it is pass through. Right?

That's at least what the commit is trying to accomplish :). I don't know the intricacies of the pass_through publications.

Reply:
Perhaps the code for the FilesystemExporter and PulpExporter should be broken up to better differentiate the two (not in this PR though).